File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [0] * (n + 1)

        def dfs(n):
            if n <= 3:
                return n
            if dp[n] != 0:
                return dp[n]

            dp[n] = dfs(n - 1) + dfs(n - 2)
            return dp[n]

        return dfs(n)

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        // [2 1 1 0 0]
        //  0 1 2 3 4
        //    c a b
        //    c = a + b
        //    b = a
        //    a = c
        int a = 1, b = 0;
        while (n-- > 0) {
            int c = a + b;
            b = a;
            a = c;
        }
        return a;
        //            0
        //          /   \
        //         1     2
        //       /   \
        //      2     3 
        //     / \
        //    3   4 
        
    }
}